database/migrations/2024_06_24_125455_data_migrate_caseloads_to_segments_table.php

```diff
@@ -0,0 +1,91 @@
+      Canyon GBS LLC, and we are committed to enforcing and protecting our trademarks
+      vigorously.
+    - The software solution, including services, infrastructure, and code, is offered as a
+      Software as a Service (SaaS) by Canyon GBS LLC.
+    - Use of this software implies agreement to the license terms and conditions as stated
+      in the Elastic License 2.0.
+
+    For more information or inquiries please visit our website at
+    https://www.canyongbs.com or contact us via email at legal@canyongbs.com.
+
+</COPYRIGHT>
+*/
+
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+use AdvisingApp\Segment\Models\Segment;
+use Illuminate\Database\Migrations\Migration;
+use AdvisingApp\Segment\Models\SegmentSubject;
+use AdvisingApp\CaseloadManagement\Models\Caseload;
+
+return new class () extends Migration {
+    public function up(): void
+    {
+        Caseload::query()
+            ->withTrashed()
+            ->cursor()
+            ->each(function (Caseload $caseload) {
+                $segment = new Segment();
+                $segment->name = $caseload->name;
+                $segment->description = $caseload->description;
+                $segment->filters = $caseload->filters;
+                $segment->model = $caseload->model;
+                $segment->type = $caseload->type;
+                $segment->user_id = $caseload->user_id;
+                $segment->created_at = $caseload->created_at;
+                $segment->updated_at = $caseload->updated_at;
+                $segment->deleted_at = $caseload->deleted_at;
+                $segment->save();
+
+                if ($caseload->subjects()->withTrashed()->exists()) {
+                    foreach ($caseload->subjects()->withTrashed()->get() as $caseloadSubject) {
+                        $segmentSubject = new SegmentSubject();
+                        $segmentSubject->subject_id = $caseloadSubject->subject_id;
+                        $segmentSubject->subject_type = $caseloadSubject->subject_type;
+                        $segmentSubject->segment_id = $segment->id;
+                        $segmentSubject->created_at = $caseloadSubject->created_at;
+                        $segmentSubject->updated_at = $caseloadSubject->updated_at;
+                        $segmentSubject->deleted_at = $caseloadSubject->deleted_at;
+                        $segmentSubject->save();
+                    }
+                }
+
+                if ($caseload->campaigns()->withTrashed()->exists()) {
+                    foreach ($caseload->campaigns()->withTrashed()->get() as $campaign) {
+                        $campaign->segment()->associate($segment);
+                        $campaign->save();
+                    }
+                }
+            });
+    }
+
+    public function down(): void
+    {
+        Schema::dropIfExists('segments');
+        Schema::dropIfExists('segment_subjects');
```

Comment:
We shouldn't drop these tables here should we? I think we should probably truncate them.

Reply:
Good catch - will update.